go/ct/spc/specification.go

```diff
@@ -1659,15 +1659,18 @@ func getAllRules() []Rule {
 	// --- SELFDESTRUCT ---
 
 	for revision := tosca.R07_Istanbul; revision <= NewestSupportedRevision; revision++ {
-		for _, beneficiaryAccountEmpty := range []bool{true, false} {
-			for _, beneficiaryAccountIsWarm := range []bool{true, false} {
-				for _, hasSelfDestructed := range []bool{true, false} {
-					rules = append(rules, makeSelfDestructRules(
-						revision,
-						hasSelfDestructed,
-						beneficiaryAccountEmpty,
-						beneficiaryAccountIsWarm,
-					)...)
+		for _, originatorHasFunds := range []bool{true, false} {
+			for _, beneficiaryAccountEmpty := range []bool{true, false} {
+				for _, beneficiaryAccountIsWarm := range []bool{true, false} {
+					for _, hasSelfDestructed := range []bool{true, false} {
```

Comment:
Has self-destructed is a term I am not familiar with: I think we may have introduced it in our side, and may induce confusion. The way I understand existence/emptiness is that empty accounts will be pruned from the database when a transaction is committed (block processed) 
Therefore, any self-destructed account is empty for the duration of the block execution, although exists would return true.

Therefore, this test is checking twice for empty, but not for exists.

Reply:
I have been reading on this. Yes, self-destructed is neither exists nor empty. it is its own semantic.

Reply:
It is possible that due to nested calls the same account is self-destructed multiple times during a single transaction. Something like:
- A calls B
- B self-destructs, the account of B is scheduled to be removed at the end of the transaction, and control goes back to A
- A calls B again, and B self-destructs again

This is all Ok, and the interpreters would not care to inform the DB multiple times about the intent of B to self-destruct itself. The one exception, though, are gas refunds.

To incentivise contracts to clean up unused accounts, you get a gas refund in such cases. But: you should only get this refund once. Calling it multiple times should only give you the reward once. And thus, we need to track in the state whether it is the first time the self-destruct is called or a second or third time. The effect is [here](), either releasing a refund or not.

You do not see this in the python spec for Cancun, since this refund was removed in London. In the Berlin specs you can see it [here](https://github.com/ethereum/execution-specs/blob/3282ee5bb88bdd56aabf2461360277ab8eac4032/src/ethereum/berlin/vm/instructions/system.py#L493-L494).